bigtable/client/row_set.h

```diff
@@ -0,0 +1,98 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_ROW_SET_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_ROW_SET_H_
+
+#include "bigtable/client/internal/conjunction.h"
+#include "bigtable/client/row_range.h"
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Represent (possibly non-continuous) set of row keys.
+ *
+ * Cloud Bigtable can scan non-continuous sets of rows, these sets can include
+ * a mix of specific row keys and ranges as defined by `bigtable::RowRange`.
+ */
+class RowSet {
+ public:
+  /// Create an empty set.
+  RowSet() {}
+
+  RowSet(RowSet&& rhs) noexcept = default;
+  RowSet& operator=(RowSet&& rhs) noexcept = default;
+  RowSet(RowSet const& rhs) = default;
+  RowSet& operator=(RowSet const& rhs) = default;
+
+  template <typename... Arg>
+  RowSet(Arg&&... a) {
+    // Generate a better error message when the parameters do not match.
+    static_assert(internal::conjunction<IsValidAppendAllArg<Arg>...>::value,
+                  "RowSet variadic constructor arguments must be convertible "
+                  "to bigtable::RowRange or absl::string_view");
+    AppendAll(std::forward<Arg&&>(a)...);
+  }
+
+  /// Add @p range to the set.
+  void Append(RowRange range) {
+    *row_set_.add_row_ranges() = range.as_proto_move();
+  }
+
+  /// Add @p row_key to the set.
+  void Append(std::string row_key) {
```

Comment:
`Append(std::string)` and `Append(char const*)` are both made redundant by `Append(absl::string_view)`, any reason to keep them?

Reply:
AFAIK, you cannot move from `absl::string_view` into a `std::string` so you must always copy.  The two overloads allow you to eliminate copies when possible.  If you think I am optimizing too early, happy to remove.  Also reordered the functions to explain why.